### ISSUE TITLE
Convert Dask dataframe back to Pandas

### ISSUE THREAD

**Issue by norandom** [dask/dask#1651]
Hi,

is there any way to convert a dask DataFrame back to Pandas? I have some features I need, which aren't yet implemented in Dask. However I need parallel / partitioned mapping. 

```
import dask.dataframe as dd
my_dask_ df = dd.from_pandas(df, npartitions=4)
my_dask_df.map_partitions(...)
del df
```

What I am looking for is:

```
df = my_dask_df.to_pandas()
del my_dask_df
```

Best,
Marius

**Comment by TomAugspurger** [comment]
Hi Marius; Assuming your `my_dask_df` fits in memory, you should be able to do `df = my_dask_df.compute()`.

**Comment by norandom** [comment]
That makes total sense. Thank you.

**Comment by gabefair** [comment]
And what if it doesn't fit into memory?

**Comment by mrocklin** [comment]
Then it won't make sense to turn it into a Pandas dataframe, which needs to fit into memory.

You might instead consider writing it to disk in some efficient format, like Parquet.

**Comment by matanster** [comment]
`.compute()` doesn't turn it into a pandas dataframe.

**Comment by matanster** [comment]
Can we actually view the computed contents of a dask dataframe without using `compute()` or converting it back to pandas? is there any streaming mode where we can see computed rows without waiting for the entire dataframe?

**Comment by martindurant** [comment]
You can use [`get_partition`](http://dask.pydata.org/en/latest/dataframe-api.html#dask.dataframe.Series.get_partition) to compute and fetch parts of the dataframe. This isn't streaming, whole tasks get performed, that is how Dask works. You could conceivably use streaming dataframes from the Streamz project, but I'm pretty sure that's not what you are after.

**Comment by mrocklin** [comment]
Call persist, then futures_of, then as_completed.  This requires the
dask.distributed scheduler


from dask.distributed import as_completed, futures_of
client = Client(...)

for future, result in as_completed(futures_of(df.persist()),
with_results=True):
    print(result)

**Comment by matanster** [comment]
Thanks for the helpful and comprehensive answer @martindurant! 
@mrocklin not sure I got what this code will accomplish, does it imply `as_completed` would maybe deliver rows "as they arrive", or is something really different meant here?

Really many thanks guys!

**Comment by MichaelSchroter** [comment]
Hi @mrocklin, @martindurant ,

As per my understanding there are two ways of processing larger than memory data sets.
1. To break up the dataframe in to small pieces (chunks) and do the processing (out-of-core). The data could be loaded in chunks and machine learning computation could be done by sklearn's `partial_fit()` method.
2. To use the hard of computer to do the job of Random Access Memory (RAM). Since the hard disk has more capacity and at the cost of more processing time. (As in hadoop)

I have the following code where, I would like to get the information of the dask dataframe (`.info()`)
```
import dask.dataframe as dd
from dask.distributed import Client
client = Client(processes=False)

merchant = dd.read_csv('/home/michael/Elo_Merchant/merchants.csv')
new_merchant_transactions = dd.read_csv('/home/michael/Elo_Merchant/new_merchant_transactions.csv')
historical_transactions = dd.read_csv('/home/michael/Elo_Merchant/historical_transactions.csv')
train = dd.read_csv('/home/michael/Elo_Merchant/train.csv')
test = dd.read_csv('/home/michael/Elo_Merchant/test.csv')

list_ = [merchant,new_merchant_transactions,historical_transactions,train,test]

for c in list_:
    c.compute().info()
```
The above code gives me the below result with an error.

```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 334696 entries, 0 to 334695
Data columns (total 22 columns):
merchant_id                    334696 non-null object
merchant_group_id              334696 non-null int64
merchant_category_id           334696 non-null int64
subsector_id                   334696 non-null int64
numerical_1                    334696 non-null float64
numerical_2                    334696 non-null float64
category_1                     334696 non-null object
most_recent_sales_range        334696 non-null object
most_recent_purchases_range    334696 non-null object
avg_sales_lag3                 334683 non-null float64
avg_purchases_lag3             334696 non-null float64
active_months_lag3             334696 non-null int64
avg_sales_lag6                 334683 non-null float64
avg_purchases_lag6             334696 non-null float64
active_months_lag6             334696 non-null int64
avg_sales_lag12                334683 non-null float64
avg_purchases_lag12            334696 non-null float64
active_months_lag12            334696 non-null int64
category_4                     334696 non-null object
city_id                        334696 non-null int64
state_id                       334696 non-null int64
category_2                     322809 non-null float64
dtypes: float64(9), int64(8), object(5)
memory usage: 56.2+ MB
<class 'pandas.core.frame.DataFrame'>
Int64Index: 1963031 entries, 0 to 643272
Data columns (total 14 columns):
authorized_flag         object
card_id                 object
city_id                 int64
category_1              object
installments            int64
category_3              object
merchant_category_id    int64
merchant_id             object
month_lag               int64
purchase_amount         float64
purchase_date           object
category_2              float64
state_id                int64
subsector_id            int64
dtypes: float64(2), int64(6), object(6)
memory usage: 224.7+ MB
distributed.core - ERROR - 
Traceback (most recent call last):
  File "/home/michael/env/lib/python3.5/site-packages/distributed/core.py", line 346, in handle_comm
    result = yield result
  File "/home/michael/env/lib/python3.5/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 293, in result
    raise self._exception
  File "/home/michael/env/lib/python3.5/site-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
  File "/home/michael/env/lib/python3.5/site-packages/distributed/worker.py", line 932, in get_data
    data = {k: self.data[k] for k in keys if k in self.data}
  File "/home/michael/env/lib/python3.5/site-packages/distributed/worker.py", line 932, in <dictcomp>
    data = {k: self.data[k] for k in keys if k in self.data}
  File "/home/michael/env/lib/python3.5/site-packages/zict/buffer.py", line 70, in __getitem__
    return self.slow_to_fast(key)
  File "/home/michael/env/lib/python3.5/site-packages/zict/buffer.py", line 57, in slow_to_fast
    value = self.slow[key]
  File "/home/michael/env/lib/python3.5/site-packages/zict/func.py", line 39, in __getitem__
    return self.load(self.d[key])
  File "/home/michael/env/lib/python3.5/site-packages/distributed/protocol/serialize.py", line 375, in deserialize_bytes
    return deserialize(header, frames)
  File "/home/michael/env/lib/python3.5/site-packages/distributed/protocol/serialize.py", line 184, in deserialize
    return loads(header, frames)
  File "/home/michael/env/lib/python3.5/site-packages/distributed/protocol/serialize.py", line 57, in pickle_loads
    return pickle.loads(b''.join(frames))
  File "/home/michael/env/lib/python3.5/site-packages/distributed/protocol/pickle.py", line 59, in loads
    return pickle.loads(x)
MemoryError
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-34-f8b13adcd8a0> in <module>
      1 for c in list_:
----> 2     c.compute().info()

~/env/lib/python3.5/site-packages/dask/base.py in compute(self, **kwargs)
    154         dask.base.compute
    155         """
--> 156         (result,) = compute(self, traverse=False, **kwargs)
    157         return result
    158 

~/env/lib/python3.5/site-packages/dask/base.py in compute(*args, **kwargs)
    395     keys = [x.__dask_keys__() for x in collections]
    396     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 397     results = schedule(dsk, keys, **kwargs)
    398     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    399 

~/env/lib/python3.5/site-packages/distributed/client.py in get(self, dsk, keys, restrictions, loose_restrictions, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   2319             try:
   2320                 results = self.gather(packed, asynchronous=asynchronous,
-> 2321                                       direct=direct)
   2322             finally:
   2323                 for f in futures.values():

~/env/lib/python3.5/site-packages/distributed/client.py in gather(self, futures, errors, maxsize, direct, asynchronous)
   1653             return self.sync(self._gather, futures, errors=errors,
   1654                              direct=direct, local_worker=local_worker,
-> 1655                              asynchronous=asynchronous)
   1656 
   1657     @gen.coroutine

~/env/lib/python3.5/site-packages/distributed/client.py in sync(self, func, *args, **kwargs)
    671             return future
    672         else:
--> 673             return sync(self.loop, func, *args, **kwargs)
    674 
    675     def __repr__(self):

~/env/lib/python3.5/site-packages/distributed/utils.py in sync(loop, func, *args, **kwargs)
    275             e.wait(10)
    276     if error[0]:
--> 277         six.reraise(*error[0])
    278     else:
    279         return result[0]

~/env/lib/python3.5/site-packages/six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

~/env/lib/python3.5/site-packages/distributed/utils.py in f()
    260             if timeout is not None:
    261                 future = gen.with_timeout(timedelta(seconds=timeout), future)
--> 262             result[0] = yield future
    263         except Exception as exc:
    264             error[0] = sys.exc_info()

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1131 
   1132                     try:
-> 1133                         value = future.result()
   1134                     except Exception:
   1135                         self.had_exception = True

/usr/lib/python3.5/asyncio/futures.py in result(self)
    291             self._tb_logger = None
    292         if self._exception is not None:
--> 293             raise self._exception
    294         return self._result
    295 

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1139                     if exc_info is not None:
   1140                         try:
-> 1141                             yielded = self.gen.throw(*exc_info)
   1142                         finally:
   1143                             # Break up a reference to itself

~/env/lib/python3.5/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1524                 else:
   1525                     self._gather_future = future
-> 1526                 response = yield future
   1527 
   1528             if response['status'] == 'error':

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1131 
   1132                     try:
-> 1133                         value = future.result()
   1134                     except Exception:
   1135                         self.had_exception = True

/usr/lib/python3.5/asyncio/futures.py in result(self)
    291             self._tb_logger = None
    292         if self._exception is not None:
--> 293             raise self._exception
    294         return self._result
    295 

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1139                     if exc_info is not None:
   1140                         try:
-> 1141                             yielded = self.gen.throw(*exc_info)
   1142                         finally:
   1143                             # Break up a reference to itself

~/env/lib/python3.5/site-packages/distributed/client.py in _gather_remote(self, direct, local_worker)
   1564                 who_has = yield self.scheduler.who_has(keys=keys)
   1565                 data2, missing_keys, missing_workers = yield gather_from_workers(
-> 1566                     who_has, rpc=self.rpc, close=False)
   1567                 response = {'status': 'OK', 'data': data2}
   1568                 if missing_keys:

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1131 
   1132                     try:
-> 1133                         value = future.result()
   1134                     except Exception:
   1135                         self.had_exception = True

/usr/lib/python3.5/asyncio/futures.py in result(self)
    291             self._tb_logger = None
    292         if self._exception is not None:
--> 293             raise self._exception
    294         return self._result
    295 

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1139                     if exc_info is not None:
   1140                         try:
-> 1141                             yielded = self.gen.throw(*exc_info)
   1142                         finally:
   1143                             # Break up a reference to itself

~/env/lib/python3.5/site-packages/distributed/utils_comm.py in gather_from_workers(who_has, rpc, close, serializers, who)
     65             for worker, c in coroutines.items():
     66                 try:
---> 67                     r = yield c
     68                 except EnvironmentError:
     69                     missing_workers.add(worker)

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1131 
   1132                     try:
-> 1133                         value = future.result()
   1134                     except Exception:
   1135                         self.had_exception = True

/usr/lib/python3.5/asyncio/futures.py in result(self)
    291             self._tb_logger = None
    292         if self._exception is not None:
--> 293             raise self._exception
    294         return self._result
    295 

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1139                     if exc_info is not None:
   1140                         try:
-> 1141                             yielded = self.gen.throw(*exc_info)
   1142                         finally:
   1143                             # Break up a reference to itself

~/env/lib/python3.5/site-packages/distributed/worker.py in get_data_from_worker(rpc, keys, worker, who, max_connections, serializers, deserializers)
   2666                                    deserializers=deserializers,
   2667                                    op='get_data', keys=keys, who=who,
-> 2668                                    max_connections=max_connections)
   2669         try:
   2670             status = response['status']

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1131 
   1132                     try:
-> 1133                         value = future.result()
   1134                     except Exception:
   1135                         self.had_exception = True

/usr/lib/python3.5/asyncio/futures.py in result(self)
    291             self._tb_logger = None
    292         if self._exception is not None:
--> 293             raise self._exception
    294         return self._result
    295 

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1145                             exc_info = None
   1146                     else:
-> 1147                         yielded = self.gen.send(value)
   1148 
   1149                     if stack_context._state.contexts is not orig_stack_contexts:

~/env/lib/python3.5/site-packages/distributed/core.py in send_recv(comm, reply, serializers, deserializers, **kwargs)
    470     if isinstance(response, dict) and response.get('status') == 'uncaught-error':
    471         if comm.deserialize:
--> 472             six.reraise(*clean_exception(**response))
    473         else:
    474             raise Exception(response['text'])

~/env/lib/python3.5/site-packages/six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

~/env/lib/python3.5/site-packages/distributed/core.py in handle_comm(self, comm, shutting_down)
    344                         if type(result) is gen.Future:
    345                             self._ongoing_coroutines.add(result)
--> 346                             result = yield result
    347                     except (CommClosedError, CancelledError) as e:
    348                         if self.status == 'running':

~/env/lib/python3.5/site-packages/tornado/gen.py in run(self)
   1131 
   1132                     try:
-> 1133                         value = future.result()
   1134                     except Exception:
   1135                         self.had_exception = True

/usr/lib/python3.5/asyncio/futures.py in result(self)
    291             self._tb_logger = None
    292         if self._exception is not None:
--> 293             raise self._exception
    294         return self._result
    295 

~/env/lib/python3.5/site-packages/tornado/gen.py in wrapper(*args, **kwargs)
    324                 try:
    325                     orig_stack_contexts = stack_context._state.contexts
--> 326                     yielded = next(result)
    327                     if stack_context._state.contexts is not orig_stack_contexts:
    328                         yielded = _create_future()

~/env/lib/python3.5/site-packages/distributed/worker.py in get_data(self, comm, keys, who, serializers, max_connections)
    930 
    931         self.outgoing_current_count += 1
--> 932         data = {k: self.data[k] for k in keys if k in self.data}
    933 
    934         if len(data) < len(keys):

~/env/lib/python3.5/site-packages/distributed/worker.py in <dictcomp>(.0)
    930 
    931         self.outgoing_current_count += 1
--> 932         data = {k: self.data[k] for k in keys if k in self.data}
    933 
    934         if len(data) < len(keys):

~/env/lib/python3.5/site-packages/zict/buffer.py in __getitem__(self, key)
     68             return self.fast[key]
     69         elif key in self.slow:
---> 70             return self.slow_to_fast(key)
     71         else:
     72             raise KeyError(key)

~/env/lib/python3.5/site-packages/zict/buffer.py in slow_to_fast(self, key)
     55 
     56     def slow_to_fast(self, key):
---> 57         value = self.slow[key]
     58         # Avoid useless movement for heavy values
     59         if self.weight(key, value) <= self.n:

~/env/lib/python3.5/site-packages/zict/func.py in __getitem__(self, key)
     37 
     38     def __getitem__(self, key):
---> 39         return self.load(self.d[key])
     40 
     41     def __setitem__(self, key, value):

~/env/lib/python3.5/site-packages/distributed/protocol/serialize.py in deserialize_bytes(b)
    373         header = {}
    374     frames = decompress(header, frames)
--> 375     return deserialize(header, frames)
    376 
    377 

~/env/lib/python3.5/site-packages/distributed/protocol/serialize.py in deserialize(header, frames, deserializers)
    182                         "data with %s" % (name, str(list(deserializers))))
    183     dumps, loads, wants_context = families[name]
--> 184     return loads(header, frames)
    185 
    186 

~/env/lib/python3.5/site-packages/distributed/protocol/serialize.py in pickle_loads(header, frames)
     55 
     56 def pickle_loads(header, frames):
---> 57     return pickle.loads(b''.join(frames))
     58 
     59 

~/env/lib/python3.5/site-packages/distributed/protocol/pickle.py in loads(x)
     57 def loads(x):
     58     try:
---> 59         return pickle.loads(x)
     60     except Exception:
     61         logger.info("Failed to deserialize %s", x[:10000], exc_info=True)

MemoryError: 

```

Could you please correct me on my understanding if wrong and I would appreciate if you could also give me advice on how to successfully run this code. I am running this code on a single machine.The data is from kaggle.com, elo merchant competition.

Thank you

Michael

**Comment by gaursurya** [comment]
Any one know how to improve the efficiency from converting dask dataframe to pandas dataframe?
Also if I need to use Random forest I need to convert dask df to pandas df?

**Comment by mrocklin** [comment]
You can convert a dask dataframe to a pandas dataframe by calling the `.compute` method.  The actual conversion is usually pretty fast (we're just concatenating many pandas dataframes) but when calling compute you're doing many other things too because dask.dataframe is lazy.

```python
df = dd.read_csv('...')  # no work actually happens yet
df = df.compute()  # this is slow because it's doing the read_csv above
```

If you're talking about Scikit-Learn's random forest then yes, you'll need to convert your data to a Numpy array, which is the input that Scikit-Learn expects.

**Comment by MichaelSchroter** [comment]
Hi,

Please use dask only if your data **does not** fit in memory (considering you are using a single machine). If data does not fit in memory then dask would take smaller than memory pieces of data or chunks and do the computation in an iterative way. Dask does not do any computation unless some methods such as `.compute()` or `.head()` what it does is create a graph (diagram) of what needs to be done to get the final result. This may be parallel or sequential depending on the whether the computation is possible to parallelize or not. Using `.visualize()` you could see the computational graph that dask has created in order to get the desired result. 

In the case, if your data does not fit memory and you want to train the data using machine learning algorithms you could use machine learning algorithms that are trained using the stochastic gradient decent method or SGD. Such as `SGDRegressor` or `SGDClassifier` included in [dask-ml ](https://ml.dask.org/incremental.html) 

As far as I know in dask the default ml algorithm that uses SGD is a support vector machine. (SVM) 
In scikit learn you are given a few options with what machine learning algorithm you are allowed to train using [SGD](https://scikit-learn.org/stable/modules/sgd.html#sgd). You might want to check these links if you need to know if random forest regressor is included in the ml algorithms that could be trained using SGD.

Hope this helps

Thanks

Michael

**Comment by gaursurya** [comment]
Hi Guys,
Thanks for all the help!!

I am using Dask for complete my task quickly and use full use of 30-cores of machine that I am using.

Problem= Groupby dataframe using column 'Study_Visit_ID' and remove duplicates from another column 'DX_ Code'
**Attaching test example below also**
<img width="435" alt="problem dask" src="https://user-images.githubusercontent.com/38278670/53224666-2bf5a500-363b-11e9-8784-59a08ba832b1.PNG">


Using below setting for full utilization of all cores.
#####################
c = LocalCluster()  
c
#####################
Query starts from here. 
records count= ~ 1 billion
#####################
from dask import dataframe as dd 
sd = dd.from_pandas(df, npartitions=30) ##
print (sd)

def grouping(dataframe_name,col_name,uni_col_name):
    df1=pd.DataFrame(columns=dataframe_name.columns)    
    g=dataframe_name[col_name].unique().compute()
    for i in tqdm(g):
        temp=dataframe_name[dataframe_name[col_name]==i].compute()
        df1=df1.append(temp.drop_duplicates([uni_col_name],keep='first'))
    return df1
#####################
a = delayed(grouping(sd,'Study_Visit_ID','DX_CODE'))
#####################
No improvement.Please suggest !!

**Comment by gaursurya** [comment]
Hi Guys,

I got Something Magical !!!

I changed my above code and divided into Two functions

def call(dataframe_name,col_name,uni_col_name,ix):
    df1=pd.DataFrame(columns=dataframe_name.columns)
    temp=dataframe_name[dataframe_name[col_name]==ix].compute()
    df1=df1.append(temp.drop_duplicates([uni_col_name],keep='first'))

def filtering(dataframe_name,col_name,uni_col_name):
    g=dataframe_name[col_name].unique().compute()
    
    for i in g:
        df2=delayed(call)(dataframe_name,col_name,uni_col_name,i)
    return df2

b=delayed(filtering(sd,'Study_Visit_ID','DX_CODE'))

This one is really Fast but when I am using b.compute() it gives me beow error.

**AttributeError: 'DataFrame' object has no attribute 'compute'**

Now what??

**Comment by MichaelSchroter** [comment]
Hi Surya,

Looks like you could load the data into memory. However, if you are doing any other modifications the data might not fit in memory. Hence, I will just do it in dask.

```python
#load the data to dask
df = dd.read_csv('/location/of/the/file.csv') # assuming your data file is a csv. for other file types look [here](url)
#according to your output it looks like the requirement is to drop duplicates from the `Description` column. Hence:
df1 = df.drop_duplicates(subset=['Description']) #keeps the first duplicates and deletes the rest. [see](http://docs.dask.org/en/latest/dataframe-api.html?highlight=drop_duplicates#dask.dataframe.DataFrame.drop_duplicates)

df1.compute()
```

Hope this helps

**Comment by gaursurya** [comment]
Hi Michael Schroter,

Thanks for reply I pasted Image of the result I wanted.
Your Query Will remove all the duplicates from 'DX_Code' without taking care of unique values in 'Study_Visit_Id'.

Also is it better to use dask dataframe or use Pandas dataframe?

**Comment by MichaelSchroter** [comment]
Hi Surya,

Sorry you could use `df.drop_duplicates()` it will remove all duplicate rows. Your answer looks like it does not have the duplicate rows in the original data. Use this method on the dataframe and it will drop all duplicate rows.

Looks like (from your code above) you could you fit the initial data in memory. You could use `df.info()` of the pandas dataframe to check the memory usage of the data. If it comfortably fits in memory then use pandas. I say comfortably because you might need to do some joining with another tables or binarizing some variables which would make the dataframe much bigger. Furthermore, dask will make computation faster if the computations could be done in parallel. However, if you are looking at training a machine learning algorithm, in order to train it the data should fit in memory unless you use SGD training method. This method enables the the data to be trained chunk by chunk. As I have mentioned above. 

Hope this anwers your question.

Thanks

Michael

**Comment by KathyGCY** [comment]
> Hi Marius; Assuming your `my_dask_df` fits in memory, you should be able to do `df = my_dask_df.compute()`.

Hi, I tried this df.compute(), doesn't work.

> ValueError                                Traceback (most recent call last)
> <ipython-input-40-3bbdf29e8812> in <module>()
>       3 df0 = dd.read_csv('Call_RO_CVS.csv')
>       4 print("load data: " , time.clock()-t)
> ----> 5 dfpandas = df0.compute()
> 
> /Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/base.py in compute(self, **kwargs)
>      94             Extra keywords to forward to the scheduler ``get`` function.
>      95         """
> ---> 96         (result,) = compute(self, traverse=False, **kwargs)
>      97         return result
>      98 
> 
> /Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/base.py in compute(*args, **kwargs)
>     201     dsk = collections_to_dsk(variables, optimize_graph, **kwargs)
>     202     keys = [var._keys() for var in variables]
> --> 203     results = get(dsk, keys, **kwargs)
>     204 
>     205     results_iter = iter(results)
> 
> /Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/threaded.py in get(dsk, result, cache, num_workers, **kwargs)
>      74     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
>      75                         cache=cache, get_id=_thread_get_id,
> ---> 76                         **kwargs)
>      77 
>      78     # Cleanup pools associated to dead threads
> 
> /Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/async.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, raise_on_exception, rerun_exceptions_locally, callbacks, dumps, loads, **kwargs)
>     523                     _execute_task(task, data)  # Re-execute locally
>     524                 else:
> --> 525                     raise(remote_exception(res, tb))
>     526             state['cache'][key] = res
>     527             finish_task(dsk, key, state, results, keyorder.get)
> 
> ValueError: could not convert string to float: '087907-2015'
> 
> Traceback
> ---------
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/async.py", line 291, in execute_task
>     result = _execute_task(task, data)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/async.py", line 272, in _execute_task
>     return func(*args2)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/compatibility.py", line 47, in apply
>     return func(*args, **kwargs)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/dataframe/io/csv.py", line 61, in pandas_read_text
>     coerce_dtypes(df, dtypes)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/dask/dataframe/io/csv.py", line 103, in coerce_dtypes
>     df[c] = df[c].astype(dtypes[c])
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/pandas/util/_decorators.py", line 91, in wrapper
>     return func(*args, **kwargs)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/pandas/core/generic.py", line 3299, in astype
>     **kwargs)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/pandas/core/internals.py", line 3224, in astype
>     return self.apply('astype', dtype=dtype, **kwargs)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/pandas/core/internals.py", line 3091, in apply
>     applied = getattr(b, f)(**kwargs)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/pandas/core/internals.py", line 471, in astype
>     **kwargs)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/pandas/core/internals.py", line 521, in _astype
>     values = astype_nansafe(values.ravel(), dtype, copy=True)
>   File "/Users/chenyinggao/anaconda/envs/py36/lib/python3.6/site-packages/pandas/core/dtypes/cast.py", line 636, in astype_nansafe
>     return arr.astype(dtype)
>

**Comment by mrocklin** [comment]
Please provide a minimal example:
http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by NyakioMureithi** [comment]
Hi guys I am looking for a way of converting from streamz.dataframe.core.DataFrame to pandas.core.frame.DataFrame any idea?

**Comment by martindurant** [comment]
(@NyakioMureithi , this question likely belongs in the streams issues)

A streams dataframe object represents a volatile set of dataframes, i.e., it's value changes qwith time or other input. So there is no obvious way to say "get me this data", at best you could get a current snapshot of the data.

**Comment by NyakioMureithi** [comment]
@martindurant  Thanks a lot for the feedback.  I am trying to simulate streaming of geospatial data and plot it on Folium. The challenge is stream filtering seems not to be working:
This: sdf[sdf.Levels == 'Good']
Gives this error: TypeError: invalid type comparison.
any idea on how to go about this?

**Comment by CMCDragonkai** [comment]
If I'm using distributed, I want to keep the last part of my computation still in the workers. So I need to compute the dask dataframes, since the last part of the work involves using a single large pandas dataframe. Is there a way to `compute` and still stay in the distributed worker instead of passing that back to the client?

**Comment by quasiben** [comment]
@CMCDragonkai yes, you can use [persist](https://distributed.dask.org/en/latest/memory.html#persisting-collections) if you run into problems please open another issue

**Comment by mrocklin** [comment]
I suspect that you will also want to use `.repartition(npartitions=1)` if
you want to get a single pandas dataframe.  Depending on the output type
that you want (dask dataframe, future, delayed) there are a few options to
move between the three.
https://docs.dask.org/en/latest/user-interfaces.html#combining-interfaces
